### ISSUE TITLE
Pool acquire timeout equals the MCP request deadline, so every DB stall reports as blocking_dispatch_unattributed

### ISSUE THREAD

**Issue by joshmh** [Dicklesworthstone/mcp_agent_mail_rust#245]
**Product:** `mcp-agent-mail` (Rust), `am 0.3.27`
**Impact:** a 16-hour production outage was undiagnosable, not because the fault was subtle, but
because the one error that would have named it is structurally suppressed.

## The headline

`DEFAULT_POOL_TIMEOUT_MS = 30_000` (`crates/mcp-agent-mail-db/src/pool.rs:1882`) and the MCP
request deadline is also **30,000 ms** (`client_deadline_ms=30000` in the timeout diagnostics).

They are the same number. So when a tool call waits on a DB connection that never becomes
available, the pool's acquire-timeout and the request's dispatch deadline expire **in the same
instant**. The request is reported as:

```
stage=blocking_dispatch_unattributed
```

and the pool's far more specific "connection acquire timeout" **never surfaces**. Every timeout in
a 16-hour window was unattributed, and the daemon logged nothing useful, for that reason alone.

## Suggested fix (one constant)

Set the pool acquire timeout **below** the request deadline — e.g. 10s against 30s. This fixes
nothing about the underlying stall; it makes the stall **self-identifying**. Precedent already
exists in the same codebase: `crates/mcp-agent-mail-cli/src/lib.rs:47775` uses `acquire_timeout_ms:
5_000`.

Generally: **a subsystem timeout should never equal the deadline of the caller that would report
it.** Equal deadlines mask the inner error with the outer one.

## Observed behaviour, for context

Measured against a live instance (~9h uptime, healthy by CPU and memory), all reproduced with raw
`curl` against the MCP endpoint so no client library is involved. **The binary under test was an
unmodified upstream release build** — we maintain a patched fork for unrelated fixes, and we
confirmed by binary inspection that none of those patches were present in the process that
produced these numbers:

| request | result |
|---|---|
| `tools/list` (no tool execution) | 0.018s, consistent |
| `tools/call health_check` (no DB connection) | 0.52s / 1.08s / 2.19s / 6.05s — variable |
| `tools/call list_agents` (tiny project) | 30s timeout |
| `tools/call list_agents` (large project) | 30s timeout |
| `tools/call whois` | 30s timeout |
| `tools/call register_agent` | 30s timeout |

So: **calls needing a DB connection block indefinitely; calls that do not are merely slow; requests
executing no tool are instant.** The process is idle throughout — 19.6% CPU, top thread having used
113 CPU-seconds in 9h41m — so the time is spent waiting, not working. Thread sampling during a
blocked call shows the dispatch thread parked in `futex_do_wait`, no D-state, no spinning.

Independent of project size (a near-empty project times out identically to the largest), and
independent of read-vs-write (`whois` and `list_agents` are reads and hang; `send_message` is a
write and succeeded earlier).

## Two secondary observations that cost us time

**1. `archive_commit_p99` is a queue metric wearing a commit name.** It is
`storage.commit_queue_latency_us` — enqueue-to-durable **queue** latency, recorded as
`req.enqueued_at.elapsed()` at completion. Pure git work has a *separate* histogram,
`git_commit_latency_us`, which is **not** exposed in the timeout diagnostics. Two engineers
independently spent hours optimising git — benchmarking clone commit times, testing an archive
shard — off the name alone.

**2. `Log2Histogram` has no window and no decay.** One 25-second stall in 635 commits pinned the
reported p99 at 8.4s for the process's entire 2d8h lifetime, and that stale figure was displayed
beside live timeouts as though current.

Together these mean the diagnostics attached to a timeout actively point away from the fault.

### Suggested fixes for those

- Expose `git_commit_latency_us` alongside the queue metric, and rename the queue one.
- Use windowed or decaying histograms so p99 reflects current health.
- Drop `ArchiveCommit`/`ArchiveWbq` from tool-timeout diagnostics, or annotate them as
  off-request-path — since the ack-fast change, archive latency cannot cause a tool timeout, so
  showing an 8s archive p99 beside a 30s tool hang invites exactly the wrong conclusion.

## What we are NOT claiming

We have **not** identified the root cause of the connection stall itself. Pool defaults are
min 25 / max 100, and only ~10 connections were open while every DB call blocked, so simple pool
*exhaustion* looks unlikely — the wait appears to be inside acquisition rather than a queue for a
free connection. We are not asserting a leak; an earlier hypothesis of ours pointed that way and
the evidence did not support it.

We also could not obtain a stack trace: the distributed binary is stripped (`nm` yields no
symbols), so attaching a debugger produced only addresses.

**Comment by Dicklesworthstone** [comment]
Confirmed, and the headline fix is on `main` (0f897fb2). Ships in the next release.

## Verification

- `DEFAULT_POOL_TIMEOUT_MS = 30_000` and the dispatch deadline `ECOSYSTEM_CLIENT_DEADLINE_MS = 30_000` were indeed the same number, so the pool's typed acquire-timeout error could never win the race against the outer deadline — every acquire stall reported as `stage=blocking_dispatch_unattributed`, as you observed.
- This is pre-existing since the initial Rust port, not a v0.3.27 regression. Telling detail: the doc comment above the constant claimed *"Reduced from legacy 60s to 15s"* while the code said 30s — the intent that it sit below the caller's deadline was there and never enforced.
- Your stall-location inference matches the pool internals: the acquire deadline is only enforced on the *wait-for-a-free-connection* path (100ms condvar waits — consistent with your `futex_do_wait` sample), while the connection-create/init path is not bounded by it.

## The fix

`DEFAULT_POOL_TIMEOUT_MS` is now **10_000** (10s vs the 30s deadline), with a rewritten doc comment stating the invariant ("a subsystem timeout should never equal the deadline of the caller that would report it") and a regression test pinning a ≥2x margin against `ECOSYSTEM_CLIENT_DEADLINE_MS` so the relationship can't silently regress if either constant moves. Because the timeout bounds only the wait, slow first-time init (migrations, archive reconstruction) is not cut off by the lower value. `DATABASE_POOL_TIMEOUT` still overrides.

With this, a stalled acquire surfaces `acquire timeout: no connections available` at 10s as a tool error — self-identifying, 20s before the deadline — instead of an unattributed 30s dispatch blowout.

## Your secondary observations

Both are legitimate and are **not** addressed by this change; noting them as follow-ups rather than silently dropping them:

- `archive_commit_p99` is `storage.commit_queue_latency_us` (enqueue-to-durable), and `git_commit_latency_us` is not exposed in timeout diagnostics — naming/exposure should change together.
- `Log2Histogram` has no window/decay, so one historic stall pins p99 for the process lifetime next to live timeouts.

One more inconsistency found while verifying, for the record: the `DATABASE_POOL_TIMEOUT` env var is parsed as **milliseconds** by the pool's `from_env` (tool-call pools) but as **seconds** by the server readiness path (`config.database_pool_timeout * 1000`). Anyone overriding the default should be aware until that's harmonized.

We have deliberately not claimed a root cause for your underlying acquisition stall — this change makes it self-identifying, which is what this issue asked for.
